### ISSUE TITLE
RDF Deserialization test cases

### ISSUE THREAD

**Issue by clehner** [timothee-haudebourg/json-ld#8]
Hi @timothee-haudebourg,

I worked on some code implementing [JSON-LD to RDF Deserialization](https://w3c.github.io/json-ld-api/#deserialize-json-ld-to-rdf-algorithm), using your library for JSON-LD expansion. That code is now released, under Apache-2.0, at <https://github.com/spruceid/ssi/>.

While testing against the JSON-LD test suite [test cases for RDF Deserialization](https://w3c.github.io/json-ld-api/tests/toRdf-manifest.html), I found a handful of tests that I could not get to pass. I've identified them in the code here: <https://github.com/spruceid/ssi/blob/main/src/jsonld.rs#L1595>. I think some of them test functionality that is implemented in Expansion or IRI Resolution rather than in Deserialization as such. I know RDF Serialization/Deserialization is not something for this crate to handle, but I wonder if you would consider possible changes to accomodate these test cases if appropriate. I could try to rewrite the test cases as JSON-LD → JSON-LD to make them usable in this crate without RDF/NQuads. The specific test cases are the following:

- [Test tli12 List with bad @base.](https://w3c.github.io/json-ld-api/tests/toRdf-manifest.html#tli12)
- [Test te111 Various relative IRIs as properties with with relative @vocab itself relative to an existing vocabulary base](https://w3c.github.io/json-ld-api/tests/toRdf-manifest.html#te111)
- [Test te112 Various relative IRIs as properties with with relative @vocab relative to another relative vocabulary base](https://w3c.github.io/json-ld-api/tests/toRdf-manifest.html#te112)
- [Test te122 Ignore some IRIs when that start with @ when expanding.](https://w3c.github.io/json-ld-api/tests/toRdf-manifest.html#te122)
- [Test tc037 property-scoped contexts which are alias of @nest](https://w3c.github.io/json-ld-api/tests/toRdf-manifest.html#tc037)
- [Test tc038 Bibframe example (poor-mans inferrence)](https://w3c.github.io/json-ld-api/tests/toRdf-manifest.html#tc038)
- [Test t0122 IRI Resolution (2)](https://w3c.github.io/json-ld-api/tests/toRdf-manifest.html#t0122)
- [Test t0123 IRI Resolution (3)](https://w3c.github.io/json-ld-api/tests/toRdf-manifest.html#t0123)
- [Test t0124 IRI Resolution (4)](https://w3c.github.io/json-ld-api/tests/toRdf-manifest.html#t0124)
- [Test t0125 IRI Resolution (5)](https://w3c.github.io/json-ld-api/tests/toRdf-manifest.html#t0125)

Edit: Another one is  [Test twf05 Triples including invalid language tags are rejected](https://w3c.github.io/json-ld-api/tests/toRdf-manifest.html#twf05), for which I propose a fix in https://github.com/timothee-haudebourg/json-ld/pull/7

**Comment by timothee-haudebourg** [comment]
Hi,

sorry for the delay. I'm a bit busy these days, but I should find some free time soon to continue developing this crate and finally release a stable version. I'll also take time to look into this.

Btw, I appreciate any feedback on the current state of the API.

**Comment by clehner** [comment]
I think the API is good. I haven't yet made full use of it though. Minor things: I was surprised that `ExpandedDocument` is a HashSet rather than an `Object`. I was also surprised about using the `json` crate as I am already using `serde_json`. I ended up making some possibly duplicated types like for Value/List/Node-Objects. I've also really just been learning JSON-LD through this. I still think some way of passing up warnings would be useful - although the "strict mode" option covers my use case currently. Overall the crate has been working great.

**Comment by timothee-haudebourg** [comment]
>  I was surprised that `ExpandedDocument` is a `HashSet` rather than an `Object`

Well it would be too broad since the output of the expansion algorithm cannot be a value. I may change it to a `Vec` instead of a `HashSet` to drop the `Hash + PartialEq` requirement on the `Id` trait.

>  I was also surprised about using the `json` crate as I am already using `serde_json`

Yes `serde_json` seems to be more widely adopted. However I like the direction taken by the `json` crate and the future optimizations promised, although it has not been updated in quite a while. Ideally I'd like to be independent of the JSON implementation, but I don't know if this will be possible. I might switch to `serde_json` if I don't find a nice way to be independent.

> I still think some way of passing up warnings would be useful

Yes I also think it would be a nice feature, but I don't know yet what would be the best way to implement that.

**Comment by timothee-haudebourg** [comment]
> I could try to rewrite the test cases as JSON-LD → JSON-LD to make them usable in this crate without RDF/NQuads

If you can do that it would be nice. That would make more test cases and would be easier for me to fix.

**Comment by timothee-haudebourg** [comment]
After further investigations, I have found some issues in the expansion algorithm using tests `e112` and `0125`, both related to incorrect IRI reference resolutions. For the other tests, as far as I understand JSON-LD, this implementation outputs correct results.

Here are my detailed observations:

- `li12`: `@base` is not an IRI. It is rejected at step 5.7 of the [context processing algorithm](https://w3c.github.io/json-ld-api/#context-processing-algorithm) as it should be. See https://github.com/w3c/json-ld-api/issues/533.

- `e111`: Seems OK.

- `e112`: There is something weird here. IRI references are badly resolved using `@vocab`. The `@vocab` value seems to be bluntly concatenated to `@base` leading to nonsensical IRIs such as ``http://example.com/rel1./rel2#../``.

- `c037`: Seems OK.

- `c038`: Seems OK.

- `0122`: Seems OK. Note that IRIs are automatically normalized (through the `iri` crate). For instance, the object `{"@id": "urn:ex:s099", "urn:ex:p": ""},` is expanded into:
  
```json
  {
      "@id": "urn:ex:s099",
      "urn:ex:p": [
          {
              "@id": "http://a/bb/ccc/d;p?q"
          }
      ]
  }
  ```
  
The playground implementation will expand it into:
  
```json
  {
      "@id": "urn:ex:s099",
      "urn:ex:p": [
          {
              "@id": "http://a/bb/ccc/./d;p?q"
          }
      ]
  }
  ```
  
- `0123`: Same as `0122` with three new cases:

  - This implementation:
    ```json
    { "@id": "urn:ex:s133", "urn:ex:p": [{ "@id": "http://a/bb/d;p?y" }]}
    ```

    Playground implementation:
    
    ```json
    { "@id": "urn:ex:s133", "urn:ex:p": [{ "@id": "http://a/bb/ccc/../d;p?y" }]}
    ```
    
  - This implementation:
    ```json
    { "@id": "urn:ex:s135", "urn:ex:p": [{ "@id": "http://a/bb/d;p?q#s" }]}
    ```

    Playground implementation:
    
    ```json
    { "@id": "urn:ex:s135", "urn:ex:p": [{ "@id": "http://a/bb/ccc/../d;p?q#s" }]}
    ```
  - This implementation:
    ```json
    { "@id": "urn:ex:s141", "urn:ex:p": [{ "@id": "http://a/bb/d;p?q" }]}
    ```

    Playground implementation:
    
    ```json
    { "@id": "urn:ex:s141", "urn:ex:p": [{ "@id": "http://a/bb/ccc/../d;p?q" }]}
    ```

- `0124`: Same as `0122` with three new cases:

  - This implementation:
    ```json
    { "@id": "urn:ex:s175", "urn:ex:p": [{ "@id": "http://a/bb/ccc/?y" }]}
    ```

    Playground implementation:
    
    ```json
    { "@id": "urn:ex:s175", "urn:ex:p": [{ "@id": "http://a/bb/ccc/.?y" }]}
    ```
    
  - This implementation:
    ```json
    { "@id": "urn:ex:s177", "urn:ex:p": [{ "@id": "http://a/bb/ccc/#s" }]}
    ```

    Playground implementation:
    
    ```json
    { "@id": "urn:ex:s177", "urn:ex:p": [{ "@id": "http://a/bb/ccc/.#s" }]}
    ```
  - This implementation:
    ```json
    { "@id": "urn:ex:s183", "urn:ex:p": [{ "@id": "http://a/bb/ccc/" }]}
    ```

    Playground implementation:
    
    ```json
    { "@id": "urn:ex:s183", "urn:ex:p": [{ "@id": "http://a/bb/ccc/." }]}
    ```

- `0125`: Again a problem with the IRI reference resolution algorithm.

  The `@base` IRI is `http://a/bb/ccc/..`. The IRI reference `g` is resolved as `http://a/bb/g` instead of `http://a/bb/ccc/g`. This is either a problem with the `iri` crate, or related to the issue with `e112`.

@clehner do you think some of your problems are linked to the fact that IRIs are normalized?

**Comment by timothee-haudebourg** [comment]
Well actually the weird IRIs produced by `e112` seems to be expected. However the IRI `http://example.com/rel1./rel2##fragment-works` is accepted while `#` is not a valid fragment character. So this is an issue with the `iref` crate.

**Comment by timothee-haudebourg** [comment]
About `e122`: keyword-like terms are indeed ignored. In this case, this means that `@id` will be set to `null`, which in this implementation leads to an expanded empty object `{}`. The question is: is there really a semantic difference between `{ "@id": null }` and `{}`?

**Comment by timothee-haudebourg** [comment]
In `e122`, the input document is expanded into
```
[
  {
    "http://example.org/vocab/foo.bar": [
      {
        "@id": "http://example.org/@foo.bar"
      }
    ],
    "http://example.org/vocab/at": [
      {
        "@id": "http://example.org/@"
      }
    ],
    "http://example.org/vocab/ignoreme": [
      {}
    ]
  }
]
```
It is not clear to me if `{}` should be removed by the expansion algorithm or by the json-ld-to-rdf algorithm. As far as I know the only step in the expansion algorithm that allows an empty map to be dropped is step 19:
> If active property is null or `@graph`, drop free-floating values as follows:
>  - If result is a map which is empty, or contains only the entries `@value` or `@list`, set result to null.
>  - Otherwise, if result is a map whose only entry is `@id`, set result to null. When the frameExpansion flag is set, a map containing only the `@id` entry is retained.
However here, the active property is `http://example.org/vocab/ignoreme`, so I see no reason for `{}` to be dropped. I don't know enough right now about the deserialization algorithm to know if there is a way to drop `{}` there.

**Comment by timothee-haudebourg** [comment]
## Fixed test cases

The following test cases now pass. Note that even though the version 2.0.2 of `iref` is enough to have correct results,
the version 2.0.3 is required so that the expected IRIs are exactly generated. Otherwise test `0128` will not pass without comparing IRIs modulo normalization.

- `e111`: Fixed by timothee-haudebourg/iref@fbfaaa5 (See timothee-haudebourg/iref#12).
- `e112`: Fixed by timothee-haudebourg/iref@fbfaaa5 (See timothee-haudebourg/iref#12).
- `c037`: Fixed by 5cb12fd.
- `c038`: Fixed by 5cb12fd.
- `0122`: Fixed by d544b31.
- `0123`: Fixed by d544b31.
- `0124`: Fixed by d544b31.
- `e112`: Fixed by timothee-haudebourg/iref@8e97fa5 (See timothee-haudebourg/iref#13).

## Won't fix

The following test cases will probably never pass because thay are based on invalid JSON-LD documents or implementations.

- `li12`: Input is not a correct JSON-LD document. See w3c/json-ld-api#533.
- `e122`: Based on an incorrect JSON-LD expansion algorithm implementation. See w3c/json-ld-api#480.

I hope for them to be removed from the test suite, or fixed.

**Comment by clehner** [comment]
This is really great, thanks @timothee-haudebourg!
